File: Wikipedia/Code/SavedArticlesFetcher.m
@import WMF;
#import "Wikipedia-Swift.h"
#import "WMFArticleFetcher.h"
#import "MWKImageInfoFetcher.h"

NSString *const WMFArticleSaveToDiskDidFailNotification = @"WMFArticleSavedToDiskWithErrorNotification";
NSString *const WMFArticleSaveToDiskDidFailArticleURLKey = @"WMFArticleSavedToDiskWithArticleURLKey";
NSString *const WMFArticleSaveToDiskDidFailErrorKey = @"WMFArticleSavedToDiskWithErrorKey";

static DDLogLevel const WMFSavedArticlesFetcherLogLevel = DDLogLevelDebug;

#undef LOG_LEVEL_DEF
#define LOG_LEVEL_DEF WMFSavedArticlesFetcherLogLevel

NS_ASSUME_NONNULL_BEGIN

@interface SavedArticlesFetcher ()

@property (nonatomic, strong, readwrite) dispatch_queue_t accessQueue;

@property (nonatomic, strong) MWKDataStore *dataStore;
@property (nonatomic, strong) WMFArticleFetcher *articleFetcher;
@property (nonatomic, strong) WMFImageController *imageController;
@property (nonatomic, strong) MWKImageInfoFetcher *imageInfoFetcher;
@property (nonatomic, strong) WMFSavedPageSpotlightManager *spotlightManager;

@property (nonatomic, getter=isUpdating) BOOL updating;
@property (nonatomic, getter=isRunning) BOOL running;

@property (nonatomic, strong) NSMutableDictionary<NSURL *, NSURLSessionTask *> *fetchOperationsByArticleTitle;
@property (nonatomic, strong) NSMutableDictionary<NSURL *, NSError *> *errorsByArticleTitle;

@property (nonatomic, strong) NSNumber *fetchesInProcessCount;

@property (nonatomic, strong) SavedArticlesFetcherProgressManager *savedArticlesFetcherProgressManager;

@end

@implementation SavedArticlesFetcher

#pragma mark - NSObject

static SavedArticlesFetcher *_articleFetcher = nil;

- (void)dealloc {
    [self stop];
}

- (instancetype)initWithDataStore:(MWKDataStore *)dataStore
                   articleFetcher:(WMFArticleFetcher *)articleFetcher
                  imageController:(WMFImageController *)imageController
                 imageInfoFetcher:(MWKImageInfoFetcher *)imageInfoFetcher {
    NSParameterAssert(dataStore);
    NSParameterAssert(articleFetcher);
    NSParameterAssert(imageController);
    NSParameterAssert(imageInfoFetcher);
    self = [super init];
    if (self) {
        self.fetchesInProcessCount = @0;
        self.accessQueue = dispatch_queue_create("org.wikipedia.savedarticlesarticleFetcher.accessQueue", DISPATCH_QUEUE_SERIAL);
        self.fetchOperationsByArticleTitle = [NSMutableDictionary new];

        [self updateFetchesInProcessCount];

        self.errorsByArticleTitle = [NSMutableDictionary new];
        self.dataStore = dataStore;
        self.articleFetcher = articleFetcher;
        self.imageController = imageController;
        self.imageInfoFetcher = imageInfoFetcher;
        self.spotlightManager = [[WMFSavedPageSpotlightManager alloc] initWithDataStore:self.dataStore];
        self.savedArticlesFetcherProgressManager = [[SavedArticlesFetcherProgressManager alloc] initWithDelegate:self];
    }
    return self;
}

- (instancetype)initWithDataStore:(MWKDataStore *)dataStore {
    return [self initWithDataStore:dataStore
                    articleFetcher:[[WMFArticleFetcher alloc] initWithDataStore:dataStore]
                   imageController:[WMFImageController sharedInstance]
                  imageInfoFetcher:[[MWKImageInfoFetcher alloc] init]];
}

#pragma mark - Progress

// Reminder: due to the internal structure of this class and how it is presently being used, we can't simply check the 'count' of 'fetchOperationsByArticleTitle' dictionary for the total. (It doesn't reflect the actual total.) Could re-plumb this class later.
- (NSUInteger)calculateTotalArticlesToFetchCount {
    NSAssert([NSThread isMainThread], @"Must be called on the main thread");
    NSManagedObjectContext *moc = self.dataStore.viewContext;
    NSFetchRequest *request = [WMFArticle fetchRequest];
    request.includesSubentities = NO;
    request.predicate = [NSPredicate predicateWithFormat:@"savedDate != NULL && isDownloaded != YES"];
    NSError *fetchError = nil;
    NSUInteger count = [moc countForFetchRequest:request error:&fetchError];
    if (fetchError) {
        DDLogError(@"Error counting number of article to be downloaded: %@", fetchError);
    }
    return count;
}

- (void)updateFetchesInProcessCount {
    NSUInteger count = [self calculateTotalArticlesToFetchCount];
    if (count == NSNotFound) {
        return;
    }
    self.fetchesInProcessCount = @(count);
}

#pragma mark - Public

- (void)start {
    self.running = YES;
    [self observeSavedPages];
    [self update];
}

- (void)stop {
    self.running = NO;
    [self unobserveSavedPages];
}

#pragma mark - Observing

- (void)articleWasUpdated:(NSNotification *)note {
    [self update];
}

- (void)_update {
    if (self.isUpdating || !self.isRunning) {
        [self updateFetchesInProcessCount];
        return;
    }
    self.updating = YES;
    NSAssert([NSThread isMainThread], @"Update must be called on the main thread");
    NSManagedObjectContext *moc = self.dataStore.viewContext;
    
    NSFetchRequest *request = [WMFArticle fetchRequest];
    request.predicate = [NSPredicate predicateWithFormat:@"savedDate != NULL && isDownloaded != YES"];
    request.fetchLimit = 1;
    NSError *fetchError = nil;
    WMFArticle *article = [[moc executeFetchRequest:request error:&fetchError] firstObject];
    if (fetchError) {
        DDLogError(@"Error fetching next article to download: %@", fetchError);
    }
    dispatch_block_t updateAgain = ^{
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            self.updating = NO;
            [self update];
        });
    };
    if (article) {
        NSURL *articleURL = article.URL;
        if (articleURL) {
            [self fetchArticleURL:articleURL
                priority:NSURLSessionTaskPriorityLow
                failure:^(NSError *error) {
                    [self updateFetchesInProcessCount];
                    updateAgain();
                }
                success:^{
                    [self.spotlightManager addToIndexWithUrl:articleURL];
                    [self updateFetchesInProcessCount];
                    updateAgain();
                }];
        } else {
            self.updating = NO;
        }
    } else {
        NSFetchRequest *downloadedRequest = [WMFArticle fetchRequest];
        downloadedRequest.predicate = [NSPredicate predicateWithFormat:@"savedDate == nil && isDownloaded == YES"];
        downloadedRequest.fetchLimit = 1;
        NSError *downloadedFetchError = nil;
        WMFArticle *articleToDelete = [[self.dataStore.viewContext executeFetchRequest:downloadedRequest error:&downloadedFetchError] firstObject];
        if (downloadedFetchError) {
            DDLogError(@"Error fetching downloaded unsaved articles: %@", downloadedFetchError);
        }
        if (articleToDelete) {
            NSURL *articleURL = article.URL;
            if (!articleURL) {
                self.updating = NO;
                [self updateFetchesInProcessCount];
                return;
            }
            [self cancelFetchForArticleURL:articleURL];
            [self removeArticleWithURL:articleURL
                            completion:^{
                                updateAgain();
                                [self updateFetchesInProcessCount];
                            }];
            [self.spotlightManager removeFromIndexWithUrl:articleURL];
        } else {
            self.updating = NO;
            [self notifyDelegateIfFinished];
            [self updateFetchesInProcessCount];
        }
    }
}

- (void)update {
    NSAssert([NSThread isMainThread], @"Update must be called on the main thread");
    [NSObject cancelPreviousPerformRequestsWithTarget:self selector:@selector(_update) object:nil];
    [self performSelector:@selector(_update) withObject:nil afterDelay:0.5];
}

- (void)observeSavedPages {
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(articleWasUpdated:) name:WMFArticleUpdatedNotification object:nil];
}

- (void)unobserveSavedPages {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

#pragma mark - Fetch

- (void)fetchArticleURL:(NSURL *)articleURL priority:(float)priority failure:(WMFErrorHandler)failure success:(WMFSuccessHandler)success {
    WMFAssertMainThread(@"must be called on the main thread");
    if (!articleURL.wmf_title) {
        DDLogError(@"Attempted to save articleURL without title: %@", articleURL);
        failure([NSError wmf_errorWithType:WMFErrorTypeInvalidRequestParameters userInfo:nil]);
        return;
    }

    if (self.fetchOperationsByArticleTitle[articleURL]) { // Protect against duplicate fetches & infinite fetch loops
        failure([NSError wmf_errorWithType:WMFErrorTypeFetchAlreadyInProgress userInfo:nil]);
        return;
    }

    // NOTE: must check isCached to determine that all article data has been downloaded
    MWKArticle *articleFromDisk = [self.dataStore articleWithURL:articleURL];
    if (articleFromDisk.isCached) {
        // only fetch images if article was cached
        [self downloadImageDataForArticle:articleFromDisk
            failure:^(NSError *_Nonnull error) {
                dispatch_async(dispatch_get_main_queue(), ^{
                    [self didFetchArticle:articleFromDisk url:articleURL error:error];
                    failure(error);
                });
            }
            success:^{
                dispatch_async(dispatch_get_main_queue(), ^{
                    [self didFetchArticle:articleFromDisk url:articleURL error:nil];
                    success();
                });
            }];
    } else {
        self.fetchOperationsByArticleTitle[articleURL] =
            [self.articleFetcher fetchArticleForURL:articleURL
                saveToDisk:YES
                priority:priority
                progress:NULL
                failure:^(NSError *_Nonnull error) {
                    dispatch_async(dispatch_get_main_queue(), ^{
                        [self didFetchArticle:nil url:articleURL error:error];
                        failure(error);
                    });
                }
                success:^(MWKArticle *_Nonnull article) {
                    dispatch_async(self.accessQueue, ^{
                        [self downloadImageDataForArticle:article
                            failure:^(NSError *error) {
                                dispatch_async(dispatch_get_main_queue(), ^{
                                    [self didFetchArticle:article url:articleURL error:error];
                                    failure(error);
                                });
                            }
                            success:^{
                                dispatch_async(dispatch_get_main_queue(), ^{
                                    [self didFetchArticle:article url:articleURL error:nil];
                                    success();
                                });
                            }];
                    });
                }];

        [self updateFetchesInProcessCount];
    }
}

- (void)downloadImageDataForArticle:(MWKArticle *)article failure:(WMFErrorHandler)failure success:(WMFSuccessHandler)success {
    dispatch_block_t doneMigration = ^{
        [self fetchAllImagesInArticle:article
            failure:^(NSError *error) {
                failure([NSError wmf_savedPageImageDownloadError]);
            }
            success:^{
                if (success) {
                    success();
                }
            }];
    };
    if (![[NSUserDefaults wmf] wmf_didFinishLegacySavedArticleImageMigration]) {
        WMF_TECH_DEBT_TODO(This legacy migration can be removed after enough users upgrade to 5.5.0)
            [self migrateLegacyImagesInArticle:article
                                    completion:doneMigration];
    } else {
        doneMigration();
    }
}

- (void)migrateLegacyImagesInArticle:(MWKArticle *)article completion:(dispatch_block_t)completion {
    WMFImageController *imageController = [WMFImageController sharedInstance];
    NSArray<NSURL *> *legacyImageURLs = [article imageURLsForSaving];
    NSString *group = article.url.wmf_articleDatabaseKey;
    if (!group || !legacyImageURLs.count) {
        if (completion) {
            completion();
        }
        return;
    }
    [imageController migrateLegacyImageURLs:legacyImageURLs intoGroup:group completion:completion];
}

- (void)fetchAllImagesInArticle:(MWKArticle *)article failure:(WMFErrorHandler)failure success:(WMFSuccessHandler)success {
    dispatch_block_t doneMigration = ^{
        NSArray *imageURLsForSaving = [article imageURLsForSaving];
        NSString *articleKey = article.url.wmf_articleDatabaseKey;
        if (!articleKey || imageURLsForSaving.count == 0) {
            success();
            return;
        }
        [self cacheImagesForArticleKey:articleKey withURLsInBackground:imageURLsForSaving failure:failure success:success];
    };
    if (![[NSUserDefaults wmf] wmf_didFinishLegacySavedArticleImageMigration]) {
        WMF_TECH_DEBT_TODO(This legacy migration can be removed after enough users upgrade to 5.0 .5)
            [self migrateLegacyImagesInArticle:article
                                    completion:doneMigration];
    } else {
        doneMigration();
    }
}

- (void)fetchGalleryDataForArticle:(MWKArticle *)article failure:(WMFErrorHandler)failure success:(WMFSuccessHandler)success {
    WMF_TECH_DEBT_TODO(check whether on - disk image info matches what we are about to fetch)
    @weakify(self);

    [self fetchImageInfoForImagesInArticle:article
        failure:^(NSError *error) {
            failure(error);
        }
        success:^(NSArray *info) {
            @strongify(self);
            if (!self) {
                failure([NSError wmf_cancelledError]);
                return;
            }
            if (info.count == 0) {
                DDLogVerbose(@"No gallery images to fetch.");
                success();
                return;
            }

            NSArray *URLs = [info valueForKey:@"imageThumbURL"];
            [self cacheImagesForArticleKey:article.url.wmf_articleDatabaseKey withURLsInBackground:URLs failure:failure success:success];
        }];
}

- (void)fetchImageInfoForImagesInArticle:(MWKArticle *)article failure:(WMFErrorHandler)failure success:(WMFSuccessNSArrayHandler)success {
    NSArray<NSString *> *imageFileTitles =
        [MWKImage mapFilenamesFromImages:[article imagesForGallery]];

    if (imageFileTitles.count == 0) {
        DDLogVerbose(@"No image info to fetch.");
        success(imageFileTitles);
        return;
    }

    NSMutableArray *infoObjects = [NSMutableArray arrayWithCapacity:imageFileTitles.count];
    WMFTaskGroup *group = [WMFTaskGroup new];
    for (NSString *canonicalFilename in imageFileTitles) {
        [group enter];
        [self.imageInfoFetcher fetchGalleryInfoForImage:canonicalFilename
            fromSiteURL:article.url
            failure:^(NSError *_Nonnull error) {
                [group leave];
            }
            success:^(id _Nonnull object) {
                if (!object || [object isEqual:[NSNull null]]) {
                    [group leave];
                    return;
                }
                [infoObjects addObject:object];
                [group leave];
            }];
    }

    @weakify(self);
    [group waitInBackgroundAndNotifyOnQueue:self.accessQueue
                                  withBlock:^{
                                      @strongify(self);
                                      if (!self) {
                                          failure([NSError wmf_cancelledError]);
                                          return;
                                      }
                                      [self.dataStore saveImageInfo:infoObjects forArticleURL:article.url];
                                      success(infoObjects);
                                  }];
}

- (void)cacheImagesForArticleKey:(NSString *)articleKey withURLsInBackground:(NSArray<NSURL *> *)imageURLs failure:(void (^_Nonnull)(NSError *_Nonnull error))failure success:(void (^_Nonnull)(void))success {
    imageURLs = [imageURLs wmf_select:^BOOL(id obj) {
        return [obj isKindOfClass:[NSURL class]];
    }];

    if (!articleKey || [imageURLs count] == 0) {
        success();
        return;
    }

    [self.imageController permanentlyCacheInBackgroundWithUrls:imageURLs groupKey:articleKey failure:failure success:success];
}

#pragma mark - Cancellation

- (void)removeArticleWithURL:(NSURL *)URL completion:(dispatch_block_t)completion {
    [self.dataStore removeArticleWithURL:URL fromDiskWithCompletion:completion];
}

- (void)cancelFetchForArticleURL:(NSURL *)URL {
    WMFAssertMainThread(@"must be called on the main thread");
    DDLogVerbose(@"Canceling saved page download for title: %@", URL);
    [self.articleFetcher cancelFetchForArticleURL:URL];
    [self.fetchOperationsByArticleTitle removeObjectForKey:URL];
}

#pragma mark - Delegate Notification

/// Only invoke within accessQueue
- (void)didFetchArticle:(MWKArticle *__nullable)fetchedArticle
                    url:(NSURL *)url
                  error:(NSError *__nullable)error {
    WMFAssertMainThread(@"must be called on the main thread");

    //Uncomment when dropping iOS 9
    if (error) {
        // store errors for later reporting
        DDLogError(@"Failed to download saved page %@ due to error: %@", url, error);
        self.errorsByArticleTitle[url] = error;
    } else {
        DDLogInfo(@"Downloaded saved page: %@", url);
    }

    // stop tracking operation, effectively advancing the progress
    [self.fetchOperationsByArticleTitle removeObjectForKey:url];

    [self updateFetchesInProcessCount];

    WMFArticle *article = [self.dataStore fetchArticleWithURL:url];
    [article updatePropertiesForError:error];
    if (error) {
        if ([error.domain isEqualToString:NSCocoaErrorDomain] && error.code == NSFileWriteOutOfSpaceError) {
            NSDictionary *userInfo = @{WMFArticleSaveToDiskDidFailErrorKey: error, WMFArticleSaveToDiskDidFailArticleURLKey: url};
            [NSNotificationCenter.defaultCenter postNotificationName:WMFArticleSaveToDiskDidFailNotification object:nil userInfo:userInfo];
            [self stop];
            article.isDownloaded = NO;
        } else if ([error.domain isEqualToString:WMFNetworkingErrorDomain] && error.code == WMFNetworkingError_APIError && [error.userInfo[NSLocalizedFailureReasonErrorKey] isEqualToString:@"missingtitle"]) {
            article.isDownloaded = YES; // skip missing titles
        } else {
            article.isDownloaded = NO;
        }
    } else {
        article.isDownloaded = YES;
    }
    
    NSError *saveError = nil;
    [self.dataStore save:&saveError];
    if (saveError) {
        DDLogError(@"Error saving after saved articles fetch: %@", saveError);
    }
}

/// Only invoke within accessQueue
- (void)notifyDelegateIfFinished {
    //Uncomment when dropping iOS 9
    //dispatch_assert_queue_debug(self.accessQueue);
    if ([self.fetchOperationsByArticleTitle count] == 0) {
        NSError *reportedError;
        if ([self.errorsByArticleTitle count] > 0) {
            reportedError = [[self.errorsByArticleTitle allValues] firstObject];
        }

        [self.errorsByArticleTitle removeAllObjects];

        DDLogDebug(@"Finished downloading all saved pages!");

        [self finishWithError:reportedError
                  fetchedData:nil];
    }
}

@end

static NSString *const WMFSavedPageErrorDomain = @"WMFSavedPageErrorDomain";

@implementation NSError (SavedArticlesFetcherErrors)

+ (instancetype)wmf_savedPageImageDownloadError {
    return [NSError errorWithDomain:WMFSavedPageErrorDomain
                               code:1
                           userInfo:@{
                               NSLocalizedDescriptionKey: WMFLocalizedStringWithDefaultValue(@"saved-pages-image-download-error", nil, nil, @"Failed to download images for this saved page.", @"Error message shown when one or more images fails to save for offline use.")
                           }];
}

@end

NS_ASSUME_NONNULL_END

File: Wikipedia/Code/SavedArticlesFetcher.swift
import Foundation
import WMF
import CocoaLumberjackSwift

// WMFLocalizedStringWithDefaultValue(@"saved-pages-image-download-error", nil, nil, @"Failed to download images for this saved page.", @"Error message shown when one or more images fails to save for offline use.")

@objc(WMFSavedArticlesFetcher)
final class SavedArticlesFetcher: NSObject {
    @objc static let saveToDiskDidFail = NSNotification.Name("SaveToDiskDidFail")
    @objc static let saveToDiskDidFailErrorKey = "error"
    @objc static let saveToDiskDidFailArticleURLKey = "articleURL"
    
    @objc dynamic var progress: Progress = Progress()
    private var countOfFetchesInProcess: Int64 = 0 {
        didSet {
            updateProgress(with: countOfFetchesInProcess, oldValue: oldValue)
        }
    }
    
    private let dataStore: MWKDataStore
    private var backgroundTaskIdentifier: UIBackgroundTaskIdentifier = UIBackgroundTaskIdentifier.invalid
    
    private let articleCacheController: ArticleCacheController
    private let spotlightManager: WMFSavedPageSpotlightManager
    
    private var isRunning = false
    private var isUpdating = false
    
    private var currentlyFetchingArticleKeys: [String] = []
    
    @objc init?(dataStore: MWKDataStore) {
        self.dataStore = dataStore
        self.articleCacheController = dataStore.cacheController.articleCache

        spotlightManager = WMFSavedPageSpotlightManager(dataStore: dataStore)
        
        super.init()
        
        resetProgress()
        updateCountOfFetchesInProcess()
    }
    
    @objc func start() {
        self.isRunning = true
        observeSavedPages()
    }
    
    @objc func stop() {
        self.isRunning = false
        unobserveSavedPages()
    }
}

private extension SavedArticlesFetcher {
    func updateCountOfFetchesInProcess() {
        guard let count = calculateCountOfArticlesToFetch() else {
            return
        }
        countOfFetchesInProcess = count
    }
    
    func updateProgress(with newValue: Int64, oldValue: Int64) {
        progress.totalUnitCount = max(progress.totalUnitCount, newValue)
        let completedUnits = progress.totalUnitCount - newValue
        progress.completedUnitCount = completedUnits
        guard newValue == 0 else {
            return
        }
        resetProgress()
    }
    
    func resetProgress() {
        progress = Progress.discreteProgress(totalUnitCount: -1)
    }
    
    private var articlesToFetchPredicate: NSPredicate {
        let now = NSDate()
        return NSPredicate(format: "savedDate != NULL && isDownloaded != YES && (downloadRetryDate == NULL || downloadRetryDate < %@)", now)
    }
    
    func calculateCountOfArticlesToFetch() -> Int64? {
        assert(Thread.isMainThread)
        
        let moc = dataStore.viewContext
        let request = WMFArticle.fetchRequest()
        request.includesSubentities = false
        request.predicate = articlesToFetchPredicate
        
        do {
            let count = try moc.count(for: request)
            return (count >= 0) ? Int64(count) : nil
        } catch let error {
            DDLogError("Error counting number of article to be downloaded: \(error)")
            return nil
        }
    }
    
    func observeSavedPages() {
        NotificationCenter.default.addObserver(self, selector: #selector(articleWasUpdated(_:)), name: NSNotification.Name.WMFArticleUpdated, object: nil)
        // WMFArticleUpdatedNotification aren't coming through when the articles are created from a background sync, so observe syncDidFinish as well to download articles synced down from the server
        NotificationCenter.default.addObserver(self, selector: #selector(syncDidFinish), name: ReadingListsController.syncDidFinishNotification, object: nil)
    }
    
    @objc func articleWasUpdated(_ note: Notification) {
        update()
    }
    
    @objc func syncDidFinish(_ note: Notification) {
        update()
    }
    
    func unobserveSavedPages() {
        NotificationCenter.default.removeObserver(self)
    }
    
    func cancelAllRequests() {
        for articleKey in currentlyFetchingArticleKeys {
            articleCacheController.cancelTasks(groupKey: articleKey)
        }
    }
    
    func update() {
        assert(Thread.isMainThread)
        NSObject.cancelPreviousPerformRequests(withTarget: self, selector: #selector(_update), object: nil)
        perform(#selector(_update), with: nil, afterDelay: 0.5)
    }
    
    private func startBackgroundTask(expirationHandler: @escaping () -> Void) {
        guard backgroundTaskIdentifier == UIBackgroundTaskIdentifier.invalid else {
            return
        }
        
        backgroundTaskIdentifier = UIApplication.shared.beginBackgroundTask(withName: "SavedArticlesFetch", expirationHandler: { [weak self] in
            expirationHandler()
            self?.endBackgroundTask()
        })
    }
    
    private func endBackgroundTask() {
        
        guard backgroundTaskIdentifier != UIBackgroundTaskIdentifier.invalid else {
            return
        }
        
        let backgroundTaskIdentifier = self.backgroundTaskIdentifier
        self.backgroundTaskIdentifier = UIBackgroundTaskIdentifier.invalid
        UIApplication.shared.endBackgroundTask(backgroundTaskIdentifier)
    }
    
    @objc func _update() {
        if isUpdating || !isRunning {
            updateCountOfFetchesInProcess()
            return
        }
        
        isUpdating = true
        
        startBackgroundTask {
            self.cancelAllRequests()
            self.stop()
        }
        
        assert(Thread.isMainThread)
        
        let moc = dataStore.viewContext
        let request = WMFArticle.fetchRequest()
        request.predicate = articlesToFetchPredicate
        request.sortDescriptors = [NSSortDescriptor(key: "savedDate", ascending: true)]
        request.fetchLimit = 1
        
        var article: WMFArticle?
        do {
            article = try moc.fetch(request).first
        } catch let error {
            DDLogError("Error fetching next article to download: \(error)")
        }
        
        let updateAgain = {
            DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(100)) {
                self.isUpdating = false
                self.endBackgroundTask()
                self.update()
            }
        }
        
        if let articleURL = article?.url,
            let articleKey = article?.key,
            let articleObjectID = article?.objectID {
            
            articleCacheController.add(url: articleURL, groupKey: articleKey, individualCompletion: { (itemResult) in
                switch itemResult {
                case .success:
                    break
                case .failure(let error):
                    DDLogError("Failed saving an item for \(articleKey): \(error)")
                }
            }) { (groupResult) in
                DispatchQueue.main.async {
                    switch groupResult {
                    case .success(let itemKeys):
                        DDLogDebug("Successfully saved all items for \(articleKey), itemKeyCount: \(itemKeys.count)")
                        self.didFetchArticle(with: articleObjectID)
                        self.spotlightManager.addToIndex(url: articleURL as NSURL)
                        self.updateCountOfFetchesInProcess()
                    case .failure(let error):
                        DDLogError("Failed saving items for \(articleKey): \(error)")
                        self.updateCountOfFetchesInProcess()
                        self.didFailToFetchArticle(with: articleObjectID, error: error)
                    }
                    updateAgain()
                }
            }
        } else {
            let downloadedRequest = WMFArticle.fetchRequest()
            downloadedRequest.predicate = NSPredicate(format: "savedDate == NULL && isDownloaded == YES")
            downloadedRequest.sortDescriptors = [NSSortDescriptor(key: "savedDate", ascending: true)]
            downloadedRequest.fetchLimit = 1
            
            var articleToDelete: WMFArticle?
            do {
                articleToDelete = try moc.fetch(downloadedRequest).first
            } catch let error {
                DDLogError("Error fetching downloaded unsaved articles: \(error)")
            }
            
            let noArticleToDeleteCompletion = {
                self.isUpdating = false
                self.updateCountOfFetchesInProcess()
                self.endBackgroundTask()
            }
            
            if let articleToDelete = articleToDelete {
                
                guard let articleKey = articleToDelete.key,
                      let articleURL = articleToDelete.url else {
                    noArticleToDeleteCompletion()
                    return
                }
                
                let articleObjectID = articleToDelete.objectID
                
                articleCacheController.remove(groupKey: articleKey, individualCompletion: { (itemResult) in
                    switch itemResult {
                    case .success:
                        break
                    case .failure(let error):
                        DDLogError("Failed removing item for \(articleKey): \(error)")
                    }
                }) { (groupResult) in
                    DispatchQueue.main.async {
                        switch groupResult {
                        case .success:
                            DDLogDebug("Successfully removed all items for \(articleKey)")
                            self.spotlightManager.removeFromIndex(url: articleURL as NSURL)
                        case .failure(let error):
                            DDLogError("Failed removing items for \(articleKey): \(error)")
                            break
                        }
                        // Ignoring failures to ensure the DB doesn't get stuck trying
                        // to remove a cache group that doesn't exist.
                        // TODO: Clean up these DB inconsistencies in the DatabaseHousekeeper
                        self.didRemoveArticle(with: articleObjectID)
                        self.updateCountOfFetchesInProcess()
                        updateAgain()
                    }
                }
            } else {
                noArticleToDeleteCompletion()
            }
        }
    }
    
    func didFetchArticle(with managedObjectID: NSManagedObjectID) {
        operateOnArticle(with: managedObjectID) { (article) in
            article.isDownloaded = true
        }
    }
    
    func didFailToFetchArticle(with managedObjectID: NSManagedObjectID, error: Error) {
        operateOnArticle(with: managedObjectID) { (article) in
            handleFailure(with: article, error: error)
        }
    }
    
    func handleFailure(with article: WMFArticle, error: Error) {
        var underlyingError: Error = error
        if let cacheError = error as? CacheControllerError {
            switch cacheError {
            case .atLeastOneItemFailedInSync(let error):
                fallthrough
            case .atLeastOneItemFailedInFileWriter(let error):
                underlyingError = error
            default:
                break
            }
        } else if let writerError = error as? ArticleCacheDBWriterError {
            switch writerError {
            case .failureFetchingMediaList(let error):
                fallthrough
            case .failureFetchingOfflineResourceList(let error):
                underlyingError = error
            case .oneOrMoreItemsFailedToMarkDownloaded(let errors):
                underlyingError = errors.first ?? error
            default:
                break
            }
        }
        if underlyingError is RequestError {
            article.error = .apiFailed
        } else {
            let nsError = underlyingError as NSError
            if nsError.domain == NSCocoaErrorDomain && nsError.code == NSFileWriteOutOfSpaceError {
                let userInfo = [SavedArticlesFetcher.saveToDiskDidFailErrorKey: error]
                NotificationCenter.default.post(name: SavedArticlesFetcher.saveToDiskDidFail, object: self, userInfo: userInfo)
                stop()
                article.error = .saveToDiskFailed
            } else if nsError.domain == NSURLErrorDomain {
                switch nsError.code {
                case NSURLErrorTimedOut:
                    fallthrough
                case NSURLErrorCancelled:
                    fallthrough
                case NSURLErrorCannotConnectToHost:
                    fallthrough
                case NSURLErrorCannotFindHost:
                    fallthrough
                case NSURLErrorNetworkConnectionLost:
                    fallthrough
                case NSURLErrorNotConnectedToInternet:
                    stop()
                default:
                    article.error = .apiFailed
                }
            } else {
                article.error = .apiFailed
            }
        }
        let newAttemptCount =  max(1, article.downloadAttemptCount + 1)
        article.downloadAttemptCount = newAttemptCount
        let secondsFromNowToAttempt: Int64
        // pow() exists but this feels safer than converting to/from decimal, feel free to update if you know better
        switch newAttemptCount {
        case 1:
            secondsFromNowToAttempt = 30
        case 2:
            secondsFromNowToAttempt = 900
        case 3:
            secondsFromNowToAttempt = 27000
        case 4:
            secondsFromNowToAttempt = 810000
        default:
            secondsFromNowToAttempt = 2419200 // 28 days later ☣
        }
        article.downloadRetryDate = Date(timeIntervalSinceNow: TimeInterval(integerLiteral: secondsFromNowToAttempt))
    }

    func didRemoveArticle(with managedObjectID: NSManagedObjectID) {
        operateOnArticle(with: managedObjectID) { (article) in
            article.isDownloaded = false
        }
    }
    
    func operateOnArticle(with managedObjectID: NSManagedObjectID, articleBlock: (WMFArticle) -> Void) {
        guard let article = dataStore.viewContext.object(with: managedObjectID) as? WMFArticle else {
            return
        }
        articleBlock(article)
        do {
            try dataStore.save()
        } catch let error {
            DDLogError("Error saving after saved articles fetch: \(error)")
        }
    }
}
